service-workers/service-worker/registration-updateviacache.https.html

```diff
@@ -0,0 +1,230 @@
+          importTime: event.data.importTime
+        });
+      });
+
+      sw.postMessage('');
+    });
+  }
+
+  function registerViaApi(scriptUrl, opts) {
+    return navigator.serviceWorker.register(scriptUrl, opts);
+  }
+
+  async function registerViaLinkElement(scriptUrl, opts) {
+    const link = document.createElement('link');
+
+    if (link.relList.supports('serviceworker') == false) throw Error("link rel=serviceworker not supported");
+
+    link.rel = 'serviceworker';
+    link.href = scriptUrl;
+    link.scope = opts.scope;
+
+    if (opts.updateViaCache) {
+      link.updateViaCache = opts.updateViaCache;
+    }
+
+    document.head.appendChild(link);
+
+    const fullScope = new URL(opts.scope, window.location).href;
+
+    while (true) {
+      const regs = await navigator.serviceWorker.getRegistrations();
+      const reg = regs.find(r => r.scope == fullScope && (r.installing || r.waiting || r.active));
+
+      if (reg) {
+        document.head.removeChild(link);
+        return reg;
+      }
+
+      await wait(100);
+    }
+  }
+
+  async function registerViaLinkHeader(scriptUrl, opts) {
+    const link = document.createElement('link');
+
+    const fullScope = new URL(opts.scope, window.location).href;
+    scriptUrl = new URL(scriptUrl, location).href;
+
+    // Assuming that if this isn't supported, the header isn't
+    if (link.relList.supports('serviceworker') == false) throw Error("link rel=serviceworker not supported");
+
+    let linkHeader = `<${scriptUrl}>; rel=serviceworker; scope="${fullScope}"`;
+
+    if (opts.updateViaCache) {
+      linkHeader += `; updateviacache="${opts.updateViaCache}"`;
+    }
+
+    const url = new URL('resources/link-header.py', location);
+    url.searchParams.set('Link', linkHeader);
+
+    const frame = await with_iframe(url);
+
+    while (true) {
+      const regs = await navigator.serviceWorker.getRegistrations();
+      const reg = regs.find(r => r.scope == fullScope && (r.installing || r.waiting || r.active));
+
+      if (reg) {
+        frame.parentNode.removeChild(frame);
+        return reg;
+      }
+
+      await wait(100);
+    }
+  }
+
+  const registrationMethods = [
+    [registerViaApi, 'via-api'],
+    [registerViaLinkElement, 'via-link-element'],
+    [registerViaLinkHeader, 'via-link-header']
+  ];
+
+  // Testing creating registrations
+  for (const [registrationMethod, registrationMethodName] of registrationMethods) {
+    for (const updateViaCache of updateViaCacheValues) {
+      const testName = `register-${registrationMethodName}-updateViaCache-${updateViaCache}`;
+
+      promise_test(async () => {
+        await cleanup();
+
+        const opts = {scope};
+
+        if (updateViaCache) opts.updateViaCache = updateViaCache;
+
+        const reg = await registrationMethod(
+          `${scriptUrl}?test=${testName}`,
+          opts
+        );
+
+        const sw = reg.installing || reg.waiting || reg.active;
+        if (!sw) console.log(reg);
```

Comment:
Do you need this console.log()? It tends to pollute test results. Should it just be an assert() instead?